text/000-wasm-bindgen-trait-method-inheritance.md

```diff
@@ -0,0 +1,401 @@
+
+1. It adds an `AsRef<JsValue>` constraint, so `trait Foo` becomes `trait Foo: AsRef<JsValue>`
+
+2. It generates externs for each method, and rewrites the methods so that they have default implementations which call the externs.
+
+The end result is that the above two traits become translated into this:
+
+```rust
+#[wasm_bindgen]
+extern {
+    #[wasm_bindgen(method, structural, js_name = addEventListener)]
+    fn add_event_listener(this: &JsValue, name: &str, listener: &Closure<FnMut(Event)>, use_capture: bool);
+
+    #[wasm_bindgen(method, structural, js_name = removeEventListener)]
+    fn remove_event_listener(this: &JsValue, name: &str, listener: &Closure<FnMut(Event)>, use_capture: bool);
+
+    #[wasm_bindgen(method, structural, js_name = dispatchEvent)]
+    fn dispatch_event(this: &JsValue, event: Event);
+}
+
+pub trait IEventTarget: AsRef<JsValue> {
+    #[inline]
+    fn add_event_listener(&self, name: &str, listener, &Closure<FnMut(Event)>, use_capture: bool) {
+        add_event_listener(self.as_ref(), name, listener, use_capture)
+    }
+
+    #[inline]
+    fn remove_event_listener(&self, name: &str, listener: &Closure<FnMut(Event)>, use_capture: bool) {
+        remove_event_listener(self.as_ref(), name, listener, use_capture)
+    }
+
+    #[inline]
+    fn dispatch_event(&self, event: Event) {
+        dispatch_event(self.as_ref(), event)
+    }
+}
+```
+
+```rust
+#[wasm_bindgen]
+extern {
+    #[wasm_bindgen(method, structural, getter = nodeName)]
+    fn node_name(this: &JsValue) -> JsString;
+
+    #[wasm_bindgen(method, structural, getter = textContent)]
+    fn text_content(this: &JsValue) -> JsString;
+
+    #[wasm_bindgen(method, structural, js_name = appendChild)]
+    fn append_child<A: INode>(this: &JsValue, child: A) -> A;
+
+    #[wasm_bindgen(method, structural, js_name = removeChild)]
+    fn remove_child<A: INode>(this: &JsValue, child: A) -> A;
+}
+
+pub trait INode: IEventTarget {
+    #[inline]
+    fn node_name(&self) -> JsString {
+        node_name(self.as_ref())
+    }
+
+    #[inline]
+    fn text_content(&self) -> JsString {
+        text_content(self.as_ref())
+    }
+
+    #[inline]
+    fn append_child<A: INode>(&self, child: A) -> A {
+        append_child(self.as_ref(), child)
+    }
+
+    #[inline]
+    fn remove_child<A: INode>(&self, child: A) -> A {
+        remove_child(self.as_ref(), child)
+    }
+
+    // And a bunch more methods...
+}
+```
+
+As you can see, it moves the trait methods into an `extern` (which has a `wasm-bindgen` attribute).
+
+Each method that takes `&self` or `&mut self` gets translated into a `method, structural` extern.
+
+Everything else is moved as-is from the trait to the extern.
+
+Then the trait methods are rewritten so that they call the extern functions (using `as_ref()` to convert `&self` to a `&JsValue`),
+and they are marked as `#[inline]`.
+
+Now that we have defined the desired traits and methods, we can add the methods to a type like this:
```

Comment:
For now, I think just supporting WebIDL is enough, but eventually users will be writing the traits by hand (which is why I chose a convenient syntax for it).

Even though WebIDL is by far the biggest usage of class-inheritance, there are plenty of other uses of class-based inheritance in JavaScript.

If we want to instead focus solely on WebIDL for now, I can change the RFC to be more focused on that.

Reply:
As for a new attribute, I had considered that, but the default impls make it so easy to apply the traits, so I figured an attribute was overkill.

Reply:
I think it is fine to include mock syntax for what macros might look like in the future, but I think we should definitely focus on WebIDL, and make it clear that macro support is not part of this RFC, merely an example of how it could potentially be extended in the future.